### ISSUE TITLE
[Request] Replace EAX with OpenAL Soft

### ISSUE THREAD

**Issue by UserGit-2021** [CookiePLMonster/SilentPatch#45]
Implement EAX using OpenAL soft.

**Comment by CookiePLMonster** [comment]
Consider checking out [IndirectSound](https://www.indirectsound.com/).

**Comment by UserGit-2021** [comment]
> Consider checking out [IndirectSound](https://www.indirectsound.com/).

The game has an EAX 4 implementation. But indirectsound does not.

**Comment by TheComputerGuy96** [comment]
> The game has an EAX 4 implementation. But indirectsound does not.

Then [DSOAL](https://github.com/kcat/dsoal/pull/52) is what you're looking for (it currently doesn't provide official binaries though: https://github.com/kcat/dsoal/issues/13)

**Comment by UserGit-2021** [comment]
> > The game has an EAX 4 implementation. But indirectsound does not.
> 
> Then [DSOAL](https://github.com/kcat/dsoal/pull/52) is what you're looking for (it currently doesn't provide official binaries though: [kcat/dsoal#13](https://github.com/kcat/dsoal/issues/13))

For the game system shock 2 is implemented without DSOAL.
DSOAL works very crookedly

**Comment by iGom** [comment]
You can use Creative ALchemy and ALchemy Restriction Remover it works with every sound card
https://www.pcgamingwiki.com/wiki/Glossary:Sound_card#Creative

**Comment by ThreeDeeJay** [comment]
I agree this belongs more to the scope of DSOAL, which can restore EAX 4 reverb and recently [fixed a crashing issue in GTA San Andreas](https://github.com/kcat/dsoal/issues/39#issuecomment-2330314303) and binaries are available here: https://github.com/kcat/dsoal/releases

Creative ALchemy might have better compatibility but requires a Creative sound card or an unlocker as mentioned above. 
[I've even managed to use ALchemy with OpenAL Soft with some hacks.](https://github.com/kcat/dsoal/issues/51#issuecomment-2378498509) which allows fixing some things like the vertically inverted sound field in San Andreas, though III and VC require sound field *rotation*: https://github.com/kcat/openal-soft/issues/674
Also I think reVC/re3 replaced Miles Sound System with OpenAL natively but I'm not sure it really helped with EAX/reverb.

Anyhow, I know embedding DSOAL into SilentPatch would probably be a big ask (like I suggested in [MTA](https://github.com/multitheftauto/mtasa-blue/issues/2687)), but it would be nice if the devs [tested them together](https://airtable.com/shrtTsUtiAJQ7wFGI/tblNOTdmp5nHXfFGU/viwjDPuKOjBnIQunv/recocF40PTA6RrtxW) since [EAX/3D audio](https://www.youtube.com/watch?v=xxY_xaFrYm4) is an extremely underrated feature, not to mention intended by the game devs, that imo everyone should be using.

**Comment by UserGit-2021** [comment]
> You can use Creative ALchemy and ALchemy Restriction Remover to eestores EAX completely - EAX 4 as well, it works with every sound card https://www.pcgamingwiki.com/wiki/Glossary:Sound_card#Creative

Again, this is not a solution to the problems. First, the source code is closed.
Second, there is less customization.
Third, openal soft has an open source code, which means that they are guaranteed to work for future PCs (They will update if it stops working for future PCs.).

**Comment by Wolver32** [comment]
For III and VC, isn't [EAX Alchemy Fix 1.0](https://www.gtagarage.com/mods/show.php?id=27928) enough to load a custom dsound.dll from the game's folder? For SA I think the system dsound.dll is prioritized, and AFAIK there aren't mods similar to EAX Alchemy Fix 1.0 that change this behaviour.

I'll be honest I'm not a huge expert when it comes to anything audio-related, but with these old games running on Windows 10 / 11 even my untrained ears can perceive that something is off. GTA Vice City in particular is badly affected, in my experience some audio streams cut off when there are lots of things going on in the background (e.g. police chases at 4+ stars).

**Comment by CookiePLMonster** [comment]
> GTA Vice City in particular is badly affected, in my experience some audio streams cut off when there are lots of things going on in the background (e.g. police chases at 4+ stars).

Check out https://gtaforums.com/topic/983941-mss-dsound3d-software-emulation-64ch-restoring-games-audio/.

**Comment by ThreeDeeJay** [comment]
> For SA I think the system dsound.dll is prioritized, and AFAIK there aren't mods similar to EAX Alchemy Fix 1.0 that change this behaviour.

@Wolver32 AFAIK this should be enough: https://www.indirectsound.com/registryIssues.html
I made a script to automate the process here: https://github.com/kcat/dsoal/issues/34#issuecomment-1139239755

**Comment by ThreeDeeJay** [comment]
> > GTA Vice City in particular is badly affected, in my experience some audio streams cut off when there are lots of things going on in the background (e.g. police chases at 4+ stars).
> 
> Check out https://gtaforums.com/topic/983941-mss-dsound3d-software-emulation-64ch-restoring-games-audio/.

@CookiePLMonster Any idea if DirectSound3D/wrappers like ALchemy and DSOAL need this?
AFAIK "Software Emulation" just provide them with the audio pre-mixed so I think at best we'd get virtualized surround or stereo, lacking 3D audio and EAX. I do know those wrappers allow specifying higher channel/voice counts in the config file but I'm not sure if it would also need to be done on Miles Sound System's end.

**Comment by CookiePLMonster** [comment]
> @CookiePLMonster Any idea if DirectSound3D/wrappers like ALchemy and DSOAL need this?

I'm not sure, I don't think those use a directsound driver in III/VC.

**Comment by ThreeDeeJay** [comment]
> I'm not sure, I don't think those use a directsound driver in III/VC.

![image](https://github.com/user-attachments/assets/d52a2e60-6c31-41ab-a0e0-ecf221dfe94d)
@CookiePLMonster EAX implies DirectSound3D (or OpenAL in newer games)
Other Miles Sound System games of the time tend to support EAX/DS3D by extension, but I suspect it wasn't tested thoroughly (judging by how [both III/VC have a rotated soundfield](https://github.com/kcat/openal-soft/issues/674) even with X-Fi hardware in XP)

**Comment by mirh** [comment]
> The problem is in DLL redirection. The game communicates with DirectSound via mss32.dll (Miles Sound System library), which doesn't see ALchemy dsound.dll in the game root.

Oh lmao, after a decade I finally understood it's the same problem of https://github.com/kcat/dsoal/issues/34#issuecomment-1417967253
EAX Alchemy Fix 1.0 is going completely dumb about it, and doesn't get that EAX unified loads dlls through COM.

> Any idea if DirectSound3D/wrappers like ALchemy and DSOAL need this?

Mss is the audio renderer, so while related it cannot have anything to do with it?
Unless, I don't know, maybe you could implement some sort of memory search and hot-patching also injected through dsound.dll.. But it really really really doesn't seem like something that could find home into kcat's repos.

> AFAIK "Software Emulation" just provide them with the audio pre-mixed so I think at best we'd get virtualized surround or stereo, lacking 3D audio and EAX.

That is `mssds3ds.m3d` so I don't think there is the former concern. EAX though, I don't know (named would suggest otherwise, yet size is about the same of msseax3 from other games).

> Other Miles Sound System games of the time tend to support EAX/DS3D by extension

It wasn't exactly an extension.. even though a lot of audio is a superset of each other, they had different _drivers_ for each (it's not difficult to notice that all MSS games until 2005 had a dedicate folder with: `Mssa3d.m3d`, `Mssa3d2.m3d`, `Mssds3dh.m3d`, `Mssds3ds.m3d`, `Msseax.m3d`, `msseax3.m3d`, `Mssfast.m3d`, `Mssrsx.m3d`).
